test/puppetlabs/http/client/async_plaintext_test.clj

```diff
@@ -352,4 +356,110 @@
+           (is (= 200 status))
+           (let [instream body
+                 buf (make-array Byte/TYPE 4)
+                 _ (.read instream buf)]
+             ;; Make sure we can read a few chars off of the stream
+             (is (= "xxxx" (String. buf "UTF-8")))
+             ;; Indicate we read some chars
+             (if initial-bytes-read? (deliver initial-bytes-read? true))
+             ;; Read the rest and validate the content
+             (let [final-string (str "xxxx" (slurp instream))]
+               (is (= (str data "xxxx") final-string))))))))))
+
+(defn- clojure-streaming-socket-timeout
+  [data-size opts]
+  (testing (str "clojure streaming socket timeout: " data-size " bytes with opts " opts)
+    (let [data (apply str (repeat (- data-size 4) "x"))]
+      (try
+        (testwebserver/with-test-webserver-and-config
+         (build-content-handler data nil true) port {:shutdown-timeout-seconds 1}
+         (with-open [client (async/create-client {:connect-timeout-milliseconds 100 :socket-timeout-milliseconds 200})]
+           (let [response @(common/get
+                            client
+                            (str "http://localhost:" port "/hello")
+                            opts)
+                 {:keys [body error]} response]
+             (if (= :unbuffered-stream (:as opts))
+               (do
+                 (if error
+                   ;; If there's an error this should behave the same as :stream
+                   (is (instance? SocketTimeoutException error))
+                   ;; If there is no error, let's consume the body to get the exception
+                   (is (thrown? SocketTimeoutException (slurp body)))))
+               (do
+                 (is error)
+                 (is (instance? SocketTimeoutException error)))))))
+        (catch TimeoutException e
+          ;; Expected whenever a server-side failure is generated
+          )))))
+
+(defn- clojure-streaming-connection-error
+  [opts]
+  (testing (str "clojure streaming connect error")
+    (with-open [client (async/create-client {:connect-timeout-milliseconds 100})]
+      (let [response @(common/get
+                       client
+                       (str "http://localhost:" 12345 "/hello")
+                       opts)
+            {:keys [error]} response]
+        (is error)
+        (is (instance? ConnectException error))))))
+
+(deftest clojure-streaming
+  (testing "clojure streaming is consistent with different payload sizes and opts"
+    (testlogging/with-test-logging
+     (doseq [data-size [32 (* 1024) (* 1024 1024)]
+             decompress-body? [false true]
+             as [:unbuffered-stream :stream]]
+       (clojure-streaming-success data-size {:as as :decompress-body decompress-body?})
+       (clojure-streaming-socket-timeout data-size {:as as :decompress-body decompress-body?})
+       (clojure-streaming-connection-error {:as as :decompress-body decompress-body?})))))
```

Comment:
These tests would probably be more readable / maintainable of they were just done as explicit `testing` blocks / assertions, and IMO the repetition / verbosity would probably be worth it.  Having so many `if`s and assertions in the helper functions also makes it harder to debug.

I don't necessarily consider it a blocker for this PR, but whenever we've had stuff like this in our tests in the past it's ended up being frustrating and some poor sucker ends up ripping it out later when they try to change something.